File: java/Tutorials/Biometrics/Java/match-multiple-faces/src/main/java/com/neurotec/tutorials/biometrics/MatchMultipleFaces.java
package com.neurotec.tutorials.biometrics;

import java.io.IOException;
import java.util.EnumSet;

import com.neurotec.biometrics.NBiometricOperation;
import com.neurotec.biometrics.NBiometricStatus;
import com.neurotec.biometrics.NBiometricTask;
import com.neurotec.biometrics.NFace;
import com.neurotec.biometrics.NMatchingResult;
import com.neurotec.biometrics.NMatchingSpeed;
import com.neurotec.biometrics.NSubject;
import com.neurotec.biometrics.client.NBiometricClient;
import com.neurotec.licensing.NLicense;
import com.neurotec.tutorials.util.LibraryManager;
import com.neurotec.tutorials.util.Utils;

public final class MatchMultipleFaces {
	private static final String DESCRIPTION = "Demonstrates matching a face from reference image to multiple faces from other image.";
	private static final String NAME = "match-multiple-faces";
	private static final String VERSION = "9.0.0.0";

	private static void usage() {
		System.out.println("usage:");
		System.out.format("\t%s [reference face image] [multiple faces image]%n", NAME);
		System.out.println();
		System.out.println("\t[reference face image]  - filename of image with a single (reference) face.");
		System.out.println("\t[multiple faces image]  - filename of image with multiple faces.\n");
	}

	public static void main(String[] args) {
		final String components = "Biometrics.FaceExtraction,Biometrics.FaceMatching";

		LibraryManager.initLibraryPath();

		Utils.printTutorialHeader(DESCRIPTION, NAME, VERSION, args);

		if (args.length < 2) {
			usage();
			System.exit(1);
		}

		NBiometricClient biometricClient = null;
		NSubject referenceSubject = null;
		NSubject candidateSubject = null;
		NBiometricTask enrollTask = null;
		
		try {
			if (!NLicense.obtainComponents("/local", 5000, components)) {
				System.err.format("Could not obtain licenses for components: %s%n", components);
				System.exit(-1);
			}

			biometricClient = new NBiometricClient();

			referenceSubject = createSubject(args[0], args[0], false);
			candidateSubject = createSubject(args[1], args[1], true);

			NBiometricStatus status = biometricClient.createTemplate(referenceSubject);
			if (status != NBiometricStatus.OK) {
				System.out.format("Template creation was unsuccessful. Status: %s\n.", status);
				return;
			}

			status = biometricClient.createTemplate(candidateSubject);
			if (status != NBiometricStatus.OK) {
				System.out.format("Template creation was unsuccessful. Status: %s\n.", status);
				return;
			}

			enrollTask = biometricClient.createTask(EnumSet.of(NBiometricOperation.ENROLL), null);

			int i = 0;
			candidateSubject.setId(new Integer(i++).toString());
			enrollTask.getSubjects().add(candidateSubject);
			for (NSubject relatedSubject : candidateSubject.getRelatedSubjects()) {
				relatedSubject.setId(new Integer(i++).toString());
				enrollTask.getSubjects().add(relatedSubject);
			}

			biometricClient.performTask(enrollTask);
			if (enrollTask.getStatus() != NBiometricStatus.OK) {
				System.out.format("Enrollment was unsuccessful. Status: %s.\n", status);
				return;
			}

			biometricClient.setMatchingThreshold(48);

			biometricClient.setFacesMatchingSpeed(NMatchingSpeed.LOW);

			status = biometricClient.identify(referenceSubject);

			if (status == NBiometricStatus.OK) {
				for (NMatchingResult result : referenceSubject.getMatchingResults()) {
					System.out.format("Matched with ID: '%s' with score %d\n", result.getId(), result.getScore());
				}
			} else if (status == NBiometricStatus.MATCH_NOT_FOUND) {
				System.out.format("Match not found");
			} else {
				System.out.format("Identification failed. Status: %s.\n", status);
				System.exit(-1);
			}
		} catch (Throwable th) {
			Utils.handleError(th);
		} finally {
			try {
				NLicense.releaseComponents(components);
			} catch (IOException e) {
				e.printStackTrace();
			}
			if (enrollTask != null) enrollTask.dispose();
			if (referenceSubject != null) referenceSubject.dispose();
			if (candidateSubject != null) candidateSubject.dispose();
			if (biometricClient != null) biometricClient.dispose();
		}
	}

	private static NSubject createSubject(String fileName, String subjectId, boolean isMultipleSubjects) {
		NSubject subject = new NSubject();
		subject.setId(subjectId);
		subject.setMultipleSubjects(isMultipleSubjects);
		NFace face = new NFace();
		face.setFileName(fileName);
		subject.getFaces().add(face);
		return subject;
	}
}

File: java/Samples/Biometrics/CPP/SimpleFacesSampleWX/MatchMultipleFaces.cpp
#include "Precompiled.h"
#include "MatchMultipleFaces.h"
#include "Resources/OpenFolderIcon.xpm"
#include "LicensePanel.h"

using namespace Neurotec::Biometrics::Gui;
using namespace Neurotec::Biometrics;
using namespace Neurotec::Biometrics::Client;
using namespace Neurotec::Gui;

namespace Neurotec
{
	namespace Samples
	{
		DEFINE_EVENT_TYPE(EVT_MULTIFACEPAGE_EXTRACT_REFERENCE_COMPLETED);
		DEFINE_EVENT_TYPE(EVT_MULTIFACEPAGE_EXTRACT_MULTIFACE_COMPLETED);
		DEFINE_EVENT_TYPE(EVT_MULTIFACEPAGE_ENROLL_COMPLETED);
		DEFINE_EVENT_TYPE(EVT_MULTIFACEPAGE_IDENTIFY_COMPLETED);
		BEGIN_EVENT_TABLE(MatchMultipleFaces, wxPanel)
			EVT_BUTTON(ID_BUTTON_OPEN_REFERENCE_IMAGE, MatchMultipleFaces::OnButtonOpenReferenceImageClick)
			EVT_BUTTON(ID_BUTTON_OPEN_MULTIFACE_IMAGE, MatchMultipleFaces::OnButtonOpenMultifaceImageClick)
			EVT_COMMAND(wxID_ANY, EVT_MULTIFACEPAGE_EXTRACT_REFERENCE_COMPLETED, MatchMultipleFaces::OnExtractReferenceCompleted)
			EVT_COMMAND(wxID_ANY, EVT_MULTIFACEPAGE_EXTRACT_MULTIFACE_COMPLETED, MatchMultipleFaces::OnExtractMultifaceCompleted)
			EVT_COMMAND(wxID_ANY, EVT_MULTIFACEPAGE_ENROLL_COMPLETED, MatchMultipleFaces::OnEnrollCompleted)
			EVT_COMMAND(wxID_ANY, EVT_MULTIFACEPAGE_IDENTIFY_COMPLETED, MatchMultipleFaces::OnIdentifyCompleted)
		END_EVENT_TABLE()

		MatchMultipleFaces::MatchMultipleFaces(wxWindow* parent, NBiometricClient &biometricClient)
		: wxPanel(parent), m_biometricClient(biometricClient), m_referenceSubject(NULL), m_multipleFacesSubject(NULL), m_enrollmentTask(NULL)
		{
			CreateGUIControls();
		}

		void MatchMultipleFaces::CreateGUIControls()
		{
			wxBoxSizer* mainBox = new wxBoxSizer(wxVERTICAL);

			LicensePanel *licensePanel = new LicensePanel(this);
			licensePanel->RefreshComponentsStatus("Biometrics.FaceExtraction,Biometrics.FaceMatching",wxEmptyString);
			mainBox->Add(licensePanel, 0, wxEXPAND | wxALL, 2);

			wxBoxSizer *toolBarSizer = new wxBoxSizer(wxHORIZONTAL);
			m_buttonOpenReferenceImage = new wxButton(this, ID_BUTTON_OPEN_REFERENCE_IMAGE, "Open Reference Image");
			m_buttonOpenReferenceImage->SetBitmap(openFolderIcon_xpm);
			m_buttonOpenReferenceImage->SetToolTip("Open Reference Image.");
			toolBarSizer->Add(m_buttonOpenReferenceImage, 0, wxALL, 2);
			m_buttonOpenMultifaceImage = new wxButton(this, ID_BUTTON_OPEN_MULTIFACE_IMAGE, "Open Multiface Image");
			m_buttonOpenMultifaceImage->SetBitmap(openFolderIcon_xpm);
			m_buttonOpenMultifaceImage->SetToolTip("Open Multiface Image.");
			toolBarSizer->Add(m_buttonOpenMultifaceImage, 0, wxALL, 2);
			mainBox->Add(toolBarSizer, 0, wxALL | wxEXPAND, 2);

			wxFlexGridSizer *flexGridSizerView = new wxFlexGridSizer(3, 2, 5, 5);
			flexGridSizerView->AddGrowableCol(0);
			flexGridSizerView->AddGrowableCol(1);
			flexGridSizerView->AddGrowableRow(1);
			wxStaticText* staticTextReferanceImage = new wxStaticText(this, wxID_ANY, "Reference Image", wxDefaultPosition, wxDefaultSize);
			wxStaticText* staticTextMultiple = new wxStaticText(this, wxID_ANY, "Multiple Face Image", wxDefaultPosition, wxDefaultSize);
			staticTextReferanceImage->SetBackgroundColour(wxSystemSettings::GetColour(wxSYS_COLOUR_ACTIVECAPTION));
			staticTextMultiple->SetBackgroundColour(wxSystemSettings::GetColour(wxSYS_COLOUR_ACTIVECAPTION));
			wxNViewZoomSlider *zoomSliderReferance = new wxNViewZoomSlider(this);
			wxNViewZoomSlider *zoomSliderMultiple = new wxNViewZoomSlider(this);
			m_faceViewReference = new wxNFaceView(this);
			m_faceViewMultiFace = new wxNFaceView(this);
			m_faceViewReference->SetBackgroundColour(GetBackgroundColour());
			m_faceViewMultiFace->SetBackgroundColour(GetBackgroundColour());
			m_faceViewMultiFace->SetShowFaceConfidence(false);
			zoomSliderReferance->SetView(m_faceViewReference);
			zoomSliderMultiple->SetView(m_faceViewMultiFace);
			flexGridSizerView->Add(staticTextReferanceImage, 0, wxEXPAND);
			flexGridSizerView->Add(staticTextMultiple, 0, wxEXPAND | wxALL, 2);
			flexGridSizerView->Add(m_faceViewReference, 1, wxEXPAND | wxALL, 2);
			flexGridSizerView->Add(m_faceViewMultiFace, 1, wxEXPAND | wxALL, 2);
			flexGridSizerView->Add(zoomSliderReferance, 1, wxEXPAND | wxALL, 2);
			flexGridSizerView->Add(zoomSliderMultiple, 1, wxEXPAND | wxALL, 2);
			mainBox->Add(flexGridSizerView, 1, wxEXPAND | wxALL, 2);

			SetBackgroundColour(GetParent()->GetBackgroundColour());
			SetSizer(mainBox);
		}

		void MatchMultipleFaces::OnButtonOpenReferenceImageClick(wxCommandEvent& WXUNUSED(event))
		{
			wxFileDialog wxfdOpenFile(this, "Open Reference Image", wxEmptyString, wxEmptyString, Common::GetOpenFileFilterString(true, false), wxFD_OPEN | wxFD_FILE_MUST_EXIST);
			if (wxfdOpenFile.ShowModal() == wxID_OK)
			{
				m_faceViewMultiFace->SetFaceIds(wxArrayString());
				// Set template size (for matching medium is recommended) (optional)
				m_biometricClient.SetFacesTemplateSize(ntsMedium);
				NFace face;
				face.SetFileName(wxfdOpenFile.GetPath());
				m_referenceSubject = NSubject();
				m_referenceSubject.GetFaces().Add(face);
				m_faceViewReference->SetFace(face);
				NAsyncOperation operation = m_biometricClient.CreateTemplateAsync(m_referenceSubject);
				operation.AddCompletedCallback(&MatchMultipleFaces::OnExtractReferenceCompletedCallback, this);
			}
		}

		void MatchMultipleFaces::OnButtonOpenMultifaceImageClick(wxCommandEvent& WXUNUSED(event))
		{
			wxFileDialog wxfdOpenFile(this, "Open Multiface Image", wxEmptyString, wxEmptyString, Common::GetOpenFileFilterString(true, false), wxFD_OPEN | wxFD_FILE_MUST_EXIST);
			if (wxfdOpenFile.ShowModal() == wxID_OK)
			{
				// Set template size (for enrolling large is recommended) (optional)
				m_biometricClient.SetFacesTemplateSize(ntsLarge);
				NFace face;
				face.SetFileName(wxfdOpenFile.GetPath());
				m_multipleFacesSubject = NSubject();
				m_multipleFacesSubject.GetFaces().Add(face);
				m_faceViewMultiFace->SetFace(face);
				m_multipleFacesSubject.SetMultipleSubjects(true);
				NAsyncOperation operation = m_biometricClient.CreateTemplateAsync(m_multipleFacesSubject);
				operation.AddCompletedCallback(&MatchMultipleFaces::OnExtractMultifaceCompletedCallback, this);
			}
		}

		void MatchMultipleFaces::EnrollMultipleFaceSubject()
		{
			m_biometricClient.Clear();
			m_enrollmentTask = NBiometricTask(nboEnroll);
			m_multipleFacesSubject.SetId("firstSubject");
			m_enrollmentTask.GetSubjects().Add(m_multipleFacesSubject);
			// Enroll all faces
			for (int i = 0; i < m_multipleFacesSubject.GetRelatedSubjects().GetCount(); i++)
			{
				NSubject tmpSubject = m_multipleFacesSubject.GetRelatedSubjects()[i];
				tmpSubject.SetId(wxString::Format("relatedSubject%i", i));
				m_enrollmentTask.GetSubjects().Add(tmpSubject);
			}
			// Enroll subjects
			NAsyncOperation operation = m_biometricClient.PerformTaskAsync(m_enrollmentTask);
			operation.AddCompletedCallback(&MatchMultipleFaces::OnEnrollCompletedCallback, this);
		}

		void MatchMultipleFaces::MatchFaces()
		{
			NAsyncOperation operation = m_biometricClient.IdentifyAsync(m_referenceSubject);
			operation.AddCompletedCallback(&MatchMultipleFaces::OnIdentifyCompletedCallback, this);
		}

		void MatchMultipleFaces::OnExtractReferenceCompletedCallback(EventArgs args)
		{
			MatchMultipleFaces* panel = reinterpret_cast<MatchMultipleFaces*>(args.GetParam());
			wxCommandEvent ev(EVT_MULTIFACEPAGE_EXTRACT_REFERENCE_COMPLETED);
			ev.SetClientData(args.GetObject<NAsyncOperation>().RefHandle());
			wxPostEvent(panel, ev);
		}

		void MatchMultipleFaces::OnExtractMultifaceCompletedCallback(EventArgs args)
		{
			MatchMultipleFaces* panel = reinterpret_cast<MatchMultipleFaces*>(args.GetParam());
			wxCommandEvent ev(EVT_MULTIFACEPAGE_EXTRACT_MULTIFACE_COMPLETED);
			ev.SetClientData(args.GetObject<NAsyncOperation>().RefHandle());
			wxPostEvent(panel, ev);
		}

		void MatchMultipleFaces::OnEnrollCompletedCallback(EventArgs args)
		{
			MatchMultipleFaces* panel = reinterpret_cast<MatchMultipleFaces*>(args.GetParam());
			wxCommandEvent ev(EVT_MULTIFACEPAGE_ENROLL_COMPLETED);
			ev.SetClientData(args.GetObject<NAsyncOperation>().RefHandle());
			wxPostEvent(panel, ev);
		}

		void MatchMultipleFaces::OnIdentifyCompletedCallback(EventArgs args)
		{
			MatchMultipleFaces* panel = reinterpret_cast<MatchMultipleFaces*>(args.GetParam());
			wxCommandEvent ev(EVT_MULTIFACEPAGE_IDENTIFY_COMPLETED);
			ev.SetClientData(args.GetObject<NAsyncOperation>().RefHandle());
			wxPostEvent(panel, ev);
		}

		void MatchMultipleFaces::OnExtractReferenceCompleted(wxCommandEvent& event)
		{
			NAsyncOperation operation(static_cast<HNObject>(event.GetClientData()), true);
			if (operation.GetError().IsNull())
			{
				NBiometricStatus status = m_referenceSubject.GetStatus();
				if (status != nbsOk)
				{
					wxMessageBox("Could not extract template from reference image.", wxMessageBoxCaptionStr, wxOK | wxICON_ERROR);
					m_referenceSubject = NULL;
				}
				else if (!m_multipleFacesSubject.IsNull())
				{
					EnrollMultipleFaceSubject();
				}
			}
			else
			{
				wxExceptionDlg::Show(operation.GetError());
				m_referenceSubject = NULL;
			}
		}

		void MatchMultipleFaces::OnExtractMultifaceCompleted(wxCommandEvent& event)
		{
			NAsyncOperation operation(static_cast<HNObject>(event.GetClientData()), true);
			if (operation.GetError().IsNull())
			{
				NBiometricStatus status = m_multipleFacesSubject.GetStatus();
				if (status != nbsOk )
				{
					wxMessageBox("Could not extract template from multiple face image.", wxMessageBoxCaptionStr, wxOK | wxICON_ERROR);
					m_multipleFacesSubject = NULL;
				}
				else if (!m_referenceSubject.IsNull())
				{
					EnrollMultipleFaceSubject();
				}
			}
			else
			{
				wxExceptionDlg::Show(operation.GetError());
				m_multipleFacesSubject = NULL;
			}
		}

		void MatchMultipleFaces::OnEnrollCompleted(wxCommandEvent& event)
		{
			NAsyncOperation operation(static_cast<HNObject>(event.GetClientData()), true);
			if (operation.GetError().IsNull())
			{
				NBiometricStatus status = m_enrollmentTask.GetStatus();
				if (status != nbsOk)
				{
					wxString statusString = NEnum::ToString(NBiometricTypes::NBiometricStatusNativeTypeOf(), status);
					wxMessageBox(wxString::Format("Enroll failed: %s", statusString), wxMessageBoxCaptionStr, wxOK | wxICON_ERROR);
				}
				else
				{
					MatchFaces();
				}
			}
			else
			{
				wxExceptionDlg::Show(operation.GetError());
			}
		}

		void MatchMultipleFaces::OnIdentifyCompleted(wxCommandEvent& event)
		{
			NAsyncOperation operation(static_cast<HNObject>(event.GetClientData()), true);
			if (operation.GetError().IsNull())
			{
				NBiometricStatus status = m_referenceSubject.GetStatus();
				if (status == nbsOk || status == nbsMatchNotFound)
				{
					int multipleFacesCount = m_enrollmentTask.GetSubjects().GetCount();
					wxArrayString results;
					results.SetCount(multipleFacesCount);
					NSubject::MatchingResultCollection resultCollection = m_referenceSubject.GetMatchingResults();
					for (int i = 0; i < resultCollection.GetCount(); i++)
					{
						NMatchingResult result = resultCollection.Get(i);
						int score = result.GetScore();
						for (int j = 0; j < multipleFacesCount; j++)
						{
							if (result.GetId() == m_enrollmentTask.GetSubjects().Get(j).GetId())
							{
								results.Item(j) = wxString::Format("score: %i (match)", score);
							}
						}
					}
					for (int i = 0; i < multipleFacesCount; i++)
					{
						if (results[i] == wxEmptyString)
							results[i] = "score: 0";
					}
					m_faceViewMultiFace->SetFaceIds(results);
				}
			}
			else
			{
				wxExceptionDlg::Show(operation.GetError());
			}
		}
	}
}
